[Housekeeping] Add Sync Method for MAuthSigningHandler

lyndon-gabriel opened the pull request:
The DPSS project is a synchronous process and leverages .NET 5 capability to send Synchronous HttpClient request. The MAuthSigningHandler currently does do not handle Send method. This PR adds the functionality to sign sync requests.

prajon84 commented:
I have few questions:

1. What is DPSS project ?
2. Why .Net 5 ?
3. Can it not use async http client ?

johnduhart reviewed (changes requested):
No. We can't just change method signatures like this and then bump the major version number. 

If it's come to the point that you need to make non-async MAuth signed HTTP calls within DPSS, I believe we've made a critical error with the design of DPSS.

@dhoizner

lyndon-gabriel commented:
> I have few questions:
> 
> 1. What is DPSS project ?
*DPSS is a project to isolated custom functions execution in a single process for the RaveEDC application. Custom functions are custom integration client write use Medidata.Core.Objects class library. This library is over 15 years old and those use the task based async patterns. We cannot modify the code based to use async because it will break millions of custom functions. One of the goal of the project is refactor Medida.Core.Objects to make it .NET Core compatible without any breaking changes.
Custom functions makes outbounds request.*

> 2. Why .Net 5 ?
*.NET has API's for synchronous  HttpClient request.*
One of the reason for updating tp .NET w
> 3. Can it not use async http client ?
*No for all the reasons mentioned above.*

lyndon-gabriel commented:
> No. We can't just change method signatures like this and then bump the major version number.
> 
> If it's come to the point that you need to make non-async MAuth signed HTTP calls within DPSS, I believe we've made a critical error with the design of DPSS.
> 
> @dhoizner

As mentioned in the previous comment, this is to maintain backward compatibility with Medidata.Core.Objects, the alternative is to use .Result, which you in the pass have pointed out leads to thread starvation. 

So if not going to change the API then what is your recommendation? @johnduhart

lyndon-gabriel commented:
Also its a minor signature change which follow the proper naming conventions for Async function and bumping major version is for exactly for changes like this, I don't understand why we can do this, please elaborate? @johnduhart

danielloganking commented on src/Medidata.MAuth.Core/IMAuthCore.cs:
```diff
@@ -5,12 +5,17 @@ namespace Medidata.MAuth.Core
 {
     internal interface IMAuthCore
     {
-        Task<HttpRequestMessage> Sign(HttpRequestMessage request, MAuthSigningOptions options);
```
This is a breaking change. Why are we doing this way instead of making the new method `SignSync` (or some such).

lyndon-gabriel replied:
So the normal convention in .NET is for Async functions to be Postfix with Async. To avoid breaking changes I can rename to SignSync but it seems odd and does follow the standard naming convention. But I can change if this is the main concern.

danielloganking replied:
The normal convention is when _adding_ `async` versions of existing synchronous methods is to append Async. Because most of the standard library API was defined prior to async it has this naming convention but for inherently async operations it is quite conventional to leave off the suffix.

And either way the point is moot because all existing clients depend on this method signature. Not making breaking changes is more important than conforming to a naming convention.

danielloganking commented on src/Medidata.MAuth.Core/IMAuthCore.cs:
```diff
@@ -5,12 +5,17 @@ namespace Medidata.MAuth.Core
 
         (string mAuthHeaderKey, string mAuthTimeHeaderKey) GetHeaderKeys();
+
+#if NET5_0
+
+        HttpRequestMessage Sign(HttpRequestMessage request, MAuthSigningOptions options);
+#endif
```
If we plan to add a synchronous implementation why limit it to `.NET 5`?

lyndon-gabriel replied:
Good point. In general the synchronous in HttpClient was added in .NET 5, but the underlying functions does not need to be limited to .NET 5.

lyndon-gabriel replied:
Okay so I reviewed and confirmed it needs to target .NET 5 or higher because the `HttpRequestMessage.ReadAsStream` is only available in .NET 5 or higher

danielloganking commented on src/Medidata.MAuth.Core/MAuthCore.cs:
```diff
@@ -124,5 +130,96 @@ internal async Task<string> CalculatePayload(HttpRequestMessage request, Private
             return (Constants.MAuthHeaderKey, Constants.MAuthTimeHeaderKey);
         }
+
+#if NET5_0
```
Why are we adding these methods? Shouldn't the of the synchronous top level method just use AsyncToSync invocation so that under the hood we still have all the benefits of async and we don't have to maintain two versions of every method? Doesn't this violate DRY?

lyndon-gabriel replied:
Having a Sync function called async is consider a bad practice and leads to performance issues such as thread starvation.To avoid these issues the lower level methods need to synchronous.

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md

danielloganking replied:
Considering [this is _exactly_ what the "sync" version of `Send` Microsoft added to HttpClient does](https://github.com/ManickaP/runtime/blob/7844fbf39812bd591dc10dc0fdfbd5fc1137d507/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L464-L470), I can assure you this the proper manner for async to sync translations. Yes, it is a kludge; because you are doing an inherently async operation within a sync context, it _must_ be.

lyndon-gabriel replied:
https://github.com/dotnet/runtime/blob/db58dcdb2a3276a61ce7f12b5d20c40426362556/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L38-L39

When sync is called it returns a ValueTask which different from a Task. ValueTask will complete synchronous if the underlying method is sync.

danielloganking commented on src/Medidata.MAuth.Core/MAuthCoreV2.cs:
```diff
@@ -136,5 +137,115 @@ public async Task<byte[]> GetSignature(HttpRequestMessage request, Authenticatio
         }
+
+
+#if NET5_0
```
Same comment as on `MAuthCore`

danielloganking commented on src/Medidata.MAuth.Core/MAuthSigningHandler.cs:
```diff
@@ -40,6 +40,7 @@ public MAuthSigningHandler(MAuthSigningOptions options, HttpMessageHandler inner
             _options = options;
         }
 
+
```
Nit: Whitespace

danielloganking commented on src/Medidata.MAuth.Core/MAuthSigningHandler.cs:
```diff
@@ -59,13 +60,43 @@ public MAuthSigningHandler(MAuthSigningOptions options, HttpMessageHandler inner
+#if NET5_0
+        /// <summary>
+        /// Signs an HTTP request with the MAuth-specific authentication information and sends the request to the
+        /// inner handler to send to the server as an asynchronous operation.
```
NIT: This is not asynchronous.

danielloganking reviewed (changes requested):
Concerns about the code:

1. We should not reimplement the entire stack in sync style and instead have the top level sync implementation use an [async to sync translation](https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development#the-blocking-hack) to keep as much of the code shared as possible and indicate the sync method is _non preferred_
2. If we implement this, we should implement this in a backwards compatible way
3. If we implement this, we should not only implement it for .NET5

However, all of the above is moot because I have fundamental concerns about even adding a synchronous version of an inherently async operation. Web calls are `async` in .NET full stop (and have been for since .NET 4.5). I am familiar with the changes in .NET5 and know of nothing that would necessitate this. If you have some other restriction requiring a web call in a synchronous context isn't that a client problem not a library problem? And if so, shouldn't the async-to-sync translation layer occur in the client? What makes this necessary as an addition to the library?

lyndon-gabriel commented:
> Concerns about the code:
> 
> 1. We should not reimplement the entire stack in sync style and instead have the top level sync implementation use an [async to sync translation](https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development#the-blocking-hack) to keep as much of the code shared as possible and indicate the sync method is _non preferred_
> 2. If we implement this, we should implement this in a backwards compatible way
> 3. If we implement this, we should not only implement it for .NET5
> 
> However, all of the above is moot because I have fundamental concerns about even adding a synchronous version of an inherently async operation. Web calls are `async` in .NET full stop (and have been for since .NET 4.5). I am familiar with the changes in .NET5 and know of nothing that would necessitate this. If you have some other restriction requiring a web call in a synchronous context isn't that a client problem not a library problem? And if so, shouldn't the async-to-sync translation layer occur in the client? What makes this necessary as an addition to the library?

So these changes will be used in RaveEDC application. In general this is a legacy code based and we can change the API surface to be asynchronous because it will be incompatible/break millions of Custom Functions. Custom Functions are written by Clients.

Sync over Async has major performance issues which have already caused CIRT's in production. It's not a scalable solution, all the underlining methods needs to be truly synchronous. However the async functions has its own benefits, which we should leverage when we can.

The changes in MAuthSigning handler will only work in .NET 5 or above. To changes in the underlying methods does not need to be limited to .NET .

danielloganking commented:
> Sync over Async has major performance issues which have already caused CIRT's in production. It's not a scalable solution, all the underlining methods needs to be truly synchronous.

Considering the HttpClient method you are relying on already does this async to sync translation I don't see this argument holding water. No HttpClient call is truly synchronous. At some point it _always_ will translate from async to sync. How you do that can ameliorate some of the issues but yes, doing async things synchronously will always be more brittle than fixing the code to be `async`. I understand such a fix is not possible in a brownfield project like Rave (and I'm not suggesting you do that) but that doesn't mean we should abandon good practice in shared libraries.

lyndon-gabriel commented:
> > Sync over Async has major performance issues which have already caused CIRT's in production. It's not a scalable solution, all the underlining methods needs to be truly synchronous.
> 
> Considering the HttpClient method you are relying on already does this async to sync translation I don't see this argument holding water. No HttpClient call is truly synchronous. At some point it _always_ will translate from async to sync. How you do that can ameliorate some of the issues but yes, doing async things synchronously will always be more brittle than fixing the code to be `async`. I understand such a fix is not possible in a brownfield project like Rave (and I'm not suggesting you do that) but that doesn't mean we should abandon good practice in shared libraries.


Im not suggesting we abandon the async functionality. This PR only adds logic to support the use case for sync http calls. I can make the adjusts to function names to be post fix with Sync and leave the original code as is. If we agree on this pattern.

danielloganking commented:
> I can make the adjusts to function names to be post fix with Sync and leave the original code as is. If we agree on this pattern.

This is definitely a better path forward but there is a still a concern with code reuse. We are duplicating a _lot_ of code here and I'm honestly not sure how we get around it. I would have to spend some time with the HttpClient implementation to fully grok the pattern but it seems like a minimum implementation uses value task to propagate results and a control bit (🤮) to indicate if the underlying call should be made synchronously or asynchronously?

To be clear I'm not against adding a truly sync option but we shouldn't do it by copying everything and doubling our possibility for bugs or worse the ability of the two implementation's logic to drift apart.

lyndon-gabriel commented:
I address the comments, reverted the breaking changes. I Added the new methods with Sync postfix. Unfortunately the sync functionality needs to be limited to .NET 5 because it uses new API's add in DelegatingHandler and HttpMessage.

danielloganking commented on src/Medidata.MAuth.Core/MAuthCore.cs:
```diff
@@ -124,5 +126,94 @@ internal async Task<string> CalculatePayload(HttpRequestMessage request, Private
         }
+
+ #if NET5_0
+        public HttpRequestMessage SignSync(HttpRequestMessage request, MAuthSigningOptions options)
```
As the `async` implementation has no `await`s can we do something with `ValueTask` to avoid duplicating the code? I know we want a fully sync operation but I can't help but feel that the code duplication is a bug waiting to happen.

lyndon-gabriel replied:
So the task are generate from the .NET framework we just await. The only async part is reading the Request Body, I refactored the code to avoid the duplications

danielloganking commented on src/Medidata.MAuth.Core/MAuthCore.cs:
```diff
@@ -29,6 +30,7 @@ public Task<HttpRequestMessage> Sign(HttpRequestMessage request, MAuthSigningOpt
             });
         }
 
+
```
NIT: Whitespace

lyndon-gabriel commented:
@danielloganking all comments were addressed. Can we proceed with getting PR merged,?

danielloganking commented on src/Medidata.MAuth.Core/MAuthCore.cs:
```diff
@@ -61,17 +65,36 @@ public bool Verify(byte[] signedData, byte[] signature, string publicKey)
+        /// The request body as a byte array.
+        /// </param>
+        /// <returns>A Task object which will result the SHA512 hash of the signature when it completes.</returns>
+        internal byte[] GetSignatureInternal(HttpRequestMessage request, AuthenticationInfo authInfo, byte[] requestContents)
```
NIT: `GetSignatureInternal` is a bit clunky, is there a name that more accurately reflects the purpose of this method? Perhaps `GenerateSignature`?

danielloganking reviewed (commented):
Can we add Unit Tests to `MAuthCoreTests` and `MAuthCoreV2Tests` that prove the signature is the same for sync and async calls? Also do we need to add sync versions of the existing tests?

danielloganking commented on src/Medidata.MAuth.Core/MAuthSigningHandler.cs:
```diff
@@ -67,5 +67,34 @@ public MAuthSigningHandler(MAuthSigningOptions options, HttpMessageHandler inner
+        /// <param name="cancellationToken">A cancellation token to cancel operation.</param>
+        /// <returns>Returns <see cref="Task{HttpResponseMessage}"/>. The task object representing the asynchronous
+        /// operation.</returns>
+        protected override HttpResponseMessage Send(
```
NIT: Should we add a `<remarks>` comment that the `Async` version should be preferred?

lyndon-gabriel replied:
So users will not directly interact with this class, you just register it via dependency injection. The method invoked on the HttpClient instance determines which method on the HttpHandler will be called  sync/async is called.

dhoizner replied:
with a public class, i don't think we can guarantee that.

danielloganking commented on tests/Medidata.MAuth.Tests/MAuthAuthenticatorTests.cs:
```diff
@@ -226,6 +254,55 @@ public static async Task AuthenticateRequest_WithValidMWSV2Request_WillAuthentic
+                actual.Headers.GetFirstValueOrDefault<long>(Constants.MAuthTimeHeaderKeyV2)
+            );
+        }
+#endif
```
Can we add a `SignRequest_WithMWSValidRequest_SignsIdenticallySyncAndAsync()` and ``SignRequest_WithMWSV2ValidRequest_SignsIdenticallySyncAndAsync()`?

lyndon-gabriel replied:
There's already an identical test for the async version. I believe these are different scenario's and should be kept separate.

danielloganking replied:
The point is to confirm that there is no functional drift between the two. Yes, we should test them separately and these cases are now covered. But it would also be a big if the two methods of signing produced different signatures. We should cover such a case tests.

lyndon-gabriel replied:
Yeah so I agree but both test assert against the same static data. The signature is hard code in the JSON file and all test for both sync and async verifies it matches defined in the json. Im not sure how adding another test, to verify the same data will be helpful.

danielloganking reviewed (approved):
LGTM. I'd still like to see the nuts addressed and a equivalency test case but understand that since the tests are based on static data the is effectively tested. The desire for an explicit equivalency test is therefor also considered a nit and thus non-blocking.

@johnduhart @dhoizner please re-review and note any blockers.

lyndon-gabriel commented:
@johnduhart is there something you will like me to change or can we proceed to getting the PR merged. Thanks.

prajon84 reviewed (approved):
LGTM
